anndata/_io/specs/registry.py

```diff
@@ -17,6 +16,61 @@ class IOSpec(NamedTuple):
     encoding_version: str
 
 
+class NoSuchIO(KeyError):
+    def __init__(
+        self,
+        registry: Literal["read", "read_partial", "write"],
+        key: tuple[type, IOSpec | type | tuple[type, str], frozenset[str]],
+    ):
+        self.registry = registry
+        self.key = key
+        super().__init__(str(self))
+
+    def __str__(self) -> str:
+        return f"No such {self.registry} function registered: {self._get_msg()}."
+
+    def _get_msg(self) -> str:
+        registry: Mapping[
+            tuple[type, IOSpec | type | tuple[type, str], frozenset[str]], Callable
+        ] = getattr(_REGISTRY, self.registry)
```

Comment:
maybe, I’ll think about it

Reply:
Most errors can be thrown like `raise FooError(msg)`, and I would like to keep that API here. I would suggest that the method to format the message is split out into functions. E.g.: `raise NoSuchIO(_read_error_msg(...))`.

This would also make the write case super easy, since you don't even need a function at the moment. You could probably just replace `TypeError` with `NoSuchIO`.

Reply:
I don’t like that API at all. It makes so much more sense to have structured data in the error in most cases.

CPython has very slowly expanded that, but is still not there. An `AttributeError` should have a field that tells you the name of the Attribute that doesn’t exist!

Reply:
I split it into different code paths, but IDK if this is much clearer. certainly more verbose.